src/map/script.c

```diff
@@ -13546,6 +13572,184 @@ BUILDIN(getiteminfo)
+ * @param slot        as the slot# of the Item Option (0 to MAX_ITEM_OPTIONS-1)
+ * @return (int) value or -1 on failure.
+ */
+BUILDIN(getequipoption)
+{
+	int val = 0, equip_index = script_getnum(st, 2);
+	int opt_type = script_getnum(st, 3);
+	int slot = script_getnum(st, 4);
+	int i = -1;
+	struct map_session_data *sd = script->rid2sd(st);
+	
+	if (sd == NULL) {
+		script_pushint(st, -1);
+		ShowError("buildin_getequipoptioninfo: Player not attached!\n");
+		return false;
+	} else if (equip_index < 0) {
+		script_pushint(st, -1);
+		ShowError("buildin_getequipoptioninfo: Invalid equipment index %d provided.\n", equip_index);
+		return false;
+	} else if (slot < 0 || slot >= MAX_ITEM_OPTIONS) {
+		script_pushint(st, -1);
+		ShowError("buildin_getequipoptioninfo: Invalid option slot %d (Min: 0, Max: %d) provided.\n", slot, MAX_ITEM_OPTIONS-1);
+		return false;
+	} else if (equip_index > 0 && equip_index <= ARRAYLENGTH(script->equip))
+		i = pc->checkequip(sd,script->equip[equip_index-1]);
+	
+	if (&sd->status.inventory[i] != NULL) {
+		switch (opt_type) {
+			case IT_OPT_INDEX:
+				val = sd->status.inventory[i].option[slot].index;
+				break;
+			case IT_OPT_VALUE:
+				val = sd->status.inventory[i].option[slot].value;
+				break;
+			default:
+				ShowError("buildin_geteqiupoptioninfo: Invalid option data type %d provided.\n", opt_type);
+				script_pushint(st, -1);
+				break;
+		}
+	}
+	
+	script_pushint(st, val);
+	
+	return true;
+}
+
+/**
+ * Set an equipment's option value.
+ * *setequipoption(<equip_index>,<type>,<slot>,<value>);
+ *
+ * @param equip_index     as the inventory index of the equipment.
+ * @param type            as IT_OPT_INDEX or IT_OPT_VALUE
+ * @param slot            as the slot of the item option (0 to MAX_ITEM_OPTIONS-1)
+ * @param value           as the value of the option type.
+ *                        For IT_OPT_INDEX see "Name" in item_options.conf
+ *                        For IT_OPT_VALUE, the value of the script bonus.
+ * @return 0 on failure, 1 on success.
+ */
+BUILDIN(setequipoption)
+{
+	int equip_index = script_getnum(st, 2);
+	int opt_type = script_getnum(st, 3);
+	int slot = script_getnum(st, 4);
+	int value = script_getnum(st, 5);
+	int i = -1;
+	
+	struct map_session_data *sd = script->rid2sd(st);
+	struct item_option *ito = NULL;
+	
+	if (sd == NULL) {
+		script_pushint(st, 0);
+		ShowError("buildin_setequipoption: Player not attached!\n");
+		return false;
+	} else if (equip_index < 0) {
+		script_pushint(st, 0);
+		ShowError("buildin_setequipoption: Invalid equipment index %d provided.\n", equip_index);
+		return false;
+	} else if (slot < 0 || slot >= MAX_ITEM_OPTIONS) {
+		script_pushint(st, 0);
+		ShowError("buildin_setequipoption: Invalid option index %d (Min: 1, Max: %d) provided.\n", slot, MAX_ITEM_OPTIONS-1);
+		return false;
+	} else if (equip_index > 0 && equip_index <= ARRAYLENGTH(script->equip))
+		i = pc->checkequip(sd,script->equip[equip_index-1]);
+	
```

Comment:
Here "i" can have -1, and in next lines it will be buffer underflow
like sd->status.inventory[-1]. This is error